src/voter/mod.rs

```diff
@@ -501,139 +495,179 @@ impl<H, N, E: Environment<H, N>, GlobalIn, GlobalOut> Voter<H, N, E, GlobalIn, G
-						self.prospective_round = Some(prospective_round);
-						best_round_completable = Some(false);
+								process_catch_up_outcome.run(CatchUpProcessingOutcome::Bad(BadCatchUp::new()));
+								return Ok(())
+							}
+							map.entry(precommit.id.clone()).or_insert((false, false)).1 = true;
+						}
+
+						let (pv, pc) = map.into_iter().fold(
+							(0, 0),
+							|(mut pv, mut pc), (id, (prevoted, precommitted))| {
+								let weight = voters.info(&id).map(|i| i.weight()).unwrap_or(0);
+
+								if prevoted {
+									pv += weight;
+								}
+
+								if precommitted {
+									pc += weight;
+								}
+
+								(pv, pc)
+							},
+						);
+
+						let threshold = voters.threshold();
+						if pv < threshold || pc < threshold {
+							trace!(target: "afg",
+								"Ignoring invalid catch up, missing voter threshold"
+							);
+
+							process_catch_up_outcome.run(CatchUpProcessingOutcome::Bad(BadCatchUp::new()));
+							return Ok(())
+						}
 					}
-				},
-				Err(e) => {
-					// the `prospective_round` may fail because the commit votes
-					// haven't been validated against typical voting rules,
-					// e.g. we could start the round with a base that is too
-					// high compared to other honest voters and would fail to
-					// import their votes.
-					trace!(target: "afg", "Prospective round at {} has failed with: {}.",
-						   prospective_round.round_number(),
-						   e,
+
+					let mut round = crate::round::Round::new(crate::round::RoundParams {
+						round_number: catch_up.round_number,
+						voters: self.voters.clone(),
+						base: (catch_up.base_hash, catch_up.base_number),
+					});
+
+					// import prevotes first.
+					for crate::SignedPrevote { prevote, id, signature } in catch_up.prevotes {
+						match round.import_prevote(&*self.env, prevote, id, signature) {
+							Ok(_) => {},
+							Err(e) => {
+								trace!(target: "afg",
+									"Ignoring invalid catch up, error importing prevote: {:?}",
+									e,
+								);
+
+								process_catch_up_outcome.run(CatchUpProcessingOutcome::Bad(BadCatchUp::new()));
+								return Ok(());
+							},
+						}
+					}
+
+					// then precommits.
+					for crate::SignedPrecommit { precommit, id, signature } in catch_up.precommits {
+						match round.import_precommit(&*self.env, precommit, id, signature) {
+							Ok(_) => {},
+							Err(e) => {
+								trace!(target: "afg",
+									"Ignoring invalid catch up, error importing precommit: {:?}",
+									e,
+								);
+
+								process_catch_up_outcome.run(CatchUpProcessingOutcome::Bad(BadCatchUp::new()));
+								return Ok(());
+							},
+						}
+					}
+
+					let state = round.state();
+					if !state.completable {
+						process_catch_up_outcome.run(CatchUpProcessingOutcome::Bad(BadCatchUp::new()));
+						return Ok(())
+					}
+
+					// beyond this point, we set this round to the past and
+					// start voting in the next round.
+					let mut just_completed = VotingRound::completed(
```

Comment:
why do we need this to be backgrounded at all?

Reply:
As discussed, we might still get some votes on this round so it doesn't really hurt to keep it around.